x/application/keeper/msg_server_undelegate_from_gateway.go

```diff
@@ -49,21 +52,57 @@ func (k msgServer) UndelegateFromGateway(ctx context.Context, msg *types.MsgUnde
 	// Remove the gateway from the application's delegatee gateway public keys
 	foundApp.DelegateeGatewayAddresses = append(foundApp.DelegateeGatewayAddresses[:foundIdx], foundApp.DelegateeGatewayAddresses[foundIdx+1:]...)
 
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	currentBlock := sdkCtx.BlockHeight()
+
+	k.recordPendingUndelegation(&foundApp, msg.GatewayAddress, currentBlock)
+
 	// Update the application store with the new delegation
 	k.SetApplication(ctx, foundApp)
 	logger.Info(fmt.Sprintf("Successfully undelegated application from gateway for app: %+v", foundApp))
 
 	// Emit the application redelegation event
 	event := msg.NewRedelegationEvent()
-	logger.Info(fmt.Sprintf("Emitting application redelegation event %v", event))
-
-	sdkCtx := sdk.UnwrapSDKContext(ctx)
-
 	if err := sdkCtx.EventManager().EmitTypedEvent(event); err != nil {
 		logger.Error(fmt.Sprintf("Failed to emit application redelegation event: %v", err))
 		return nil, err
 	}
+	logger.Info(fmt.Sprintf("Emitted application redelegation event %v", event))
 
 	isSuccessful = true
 	return &types.MsgUndelegateFromGatewayResponse{}, nil
 }
+
+// recordPendingUndelegation adds the given gateway address to the application's
+// pending undelegations list.
+func (k Keeper) recordPendingUndelegation(
+	app *types.Application,
+	gatewayAddress string,
+	currentBlockHeight int64,
+) {
+	sessionEndHeight := uint64(sessionkeeper.GetSessionEndBlockHeight(currentBlockHeight))
+
+	// Create the session pending undelegations list entry for the given session
+	// end height if it doesn't already exist.
+	undelegatingGatewayListAtBlock, ok := app.PendingUndelegations[sessionEndHeight]
+	if !ok {
+		undelegatingGatewayListAtBlock = types.UndelegatingGatewayList{
+			GatewayAddresses: []string{},
+		}
+	}
+
+	// Add the gateway address to the undelegated gateways list if it's not already there.
+	if !slices.Contains(undelegatingGatewayListAtBlock.GatewayAddresses, gatewayAddress) {
+		undelegatingGatewayListAtBlock.GatewayAddresses = append(
+			undelegatingGatewayListAtBlock.GatewayAddresses,
+			gatewayAddress,
+		)
+		app.PendingUndelegations[sessionEndHeight] = undelegatingGatewayListAtBlock
+	} else {
+		k.logger.Warn(fmt.Sprintf(
+			"Application undelegating (again) from gateway it's already undelegating from with address [%s]",
```

Comment:
Is it worth printing the application address as well?

Wdyt about making it an `Info()` log instead of a `Warn()`? My thinking is that it's not something a poktrolld operator needs to be concerned about.